runtime/compiler/compile/J9SymbolReferenceTable.cpp

```diff
@@ -533,6 +557,45 @@ J9::SymbolReferenceTable::findOrCreateMethodTypeTableEntrySymbol(TR::ResolvedMet
    return symRef;
    }
 
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+TR::SymbolReference *
+J9::SymbolReferenceTable::refineInvokeCacheElementSymRefWithKnownObjectIndexForInvokeHandle(TR::ResolvedMethodSymbol * owningMethodSymbol,  TR::SymbolReference * originalSymRef, int32_t cpIndex, bool isMemberNameObject)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(fe());
+   TR::VMAccessCriticalSection invokeCacheEntry(comp());
+   TR::KnownObjectTable *knot = comp()->getOrCreateKnownObjectTable();
+   if (!knot) return originalSymRef;
+   TR_ResolvedJ9Method *owningMethod = static_cast<TR_ResolvedJ9Method*>(owningMethodSymbol->getResolvedMethod());
+   TR::KnownObjectTable::Index arrayElementKnotIndex = TR::KnownObjectTable::UNKNOWN;
+   if (!isMemberNameObject)
+      arrayElementKnotIndex = knot->getOrCreateIndex((uintptr_t) owningMethod->appendixElementRefFromInvokeHandleSideTable(cpIndex), true);
+   else
+      arrayElementKnotIndex = knot->getOrCreateIndex((uintptr_t) owningMethod->memberNameElementRefFromInvokeHandleSideTable(cpIndex), true);
+   
+   TR::SymbolReference *newRef = findOrCreateSymRefWithKnownObject(originalSymRef, arrayElementKnotIndex);
+   return newRef;
+   }
+
+
+TR::SymbolReference *
+J9::SymbolReferenceTable::refineInvokeCacheElementSymRefWithKnownObjectIndexForInvokeDynamic(TR::ResolvedMethodSymbol * owningMethodSymbol,  TR::SymbolReference * originalSymRef, int32_t callSiteIndex, bool isMemberNameObject)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(fe());
+   TR::VMAccessCriticalSection invokeCacheEntry(comp());
+   TR::KnownObjectTable *knot = comp()->getOrCreateKnownObjectTable();
+   if (!knot) return originalSymRef;
+   TR_ResolvedJ9Method *owningMethod = static_cast<TR_ResolvedJ9Method*>(owningMethodSymbol->getResolvedMethod());
+   TR::KnownObjectTable::Index arrayElementKnotIndex = TR::KnownObjectTable::UNKNOWN;
+   if (!isMemberNameObject)
+      arrayElementKnotIndex = knot->getOrCreateIndex((uintptr_t) owningMethod->appendixElementRefFromInvokeDynamicSideTable(callSiteIndex), true);
+   else
+      arrayElementKnotIndex = knot->getOrCreateIndex((uintptr_t) owningMethod->memberNameElementRefFromInvokeDynamicSideTable(callSiteIndex), true);
+
+   TR::SymbolReference *newRef = findOrCreateSymRefWithKnownObject(originalSymRef, arrayElementKnotIndex);
+   return newRef;
+   }
```

Comment:
The logic in the invokeHandle and invokeDynamic cases is 95% identical.  Can you think of a way of combining the internals into a shared function without impacting understanding?    I'm just thinking of the code footprint here...